Source/Evolve.cpp

```diff
@@ -33,24 +33,55 @@ Real compute_dt(const Geometry& geom, const Real cfl_factor, const MultiFab& sta
         dt_translation = std::min(std::min(dxi[0],dxi[1]), dxi[2]) / PhysConst::c * cfl_factor;
     }
 
-    Real dt_si_matter = 0.0;
+    Real dt_flavor = 0.0;
     if (flavor_cfl_factor > 0.0) {
-        // self-interaction and matter part of timestep limit
-        // NOTE: these currently over-estimate both potentials, but avoid reduction over all particles
-        // NOTE: the vacuum potential is currently ignored. This requires a min reduction over particle energies
-        Real N_diag_max = 0;
-        #include "generated_files/Evolve.cpp_compute_dt_fill"
-        Real Vmax = std::sqrt(2.) * PhysConst::GF * std::max(N_diag_max/cell_volume, state.max(GIdx::rho)/PhysConst::Mp);
-        if(Vmax>0) dt_si_matter = PhysConst::hbar/Vmax*flavor_cfl_factor;
+        // define the reduction operator to get the max contribution to
+        // the potential from matter and neutrinos
+        // compute "effective" potential (ergs) that produces characteristic timescale
+        // when multiplied by hbar
+        ReduceOps<ReduceOpMax,ReduceOpMax> reduce_op;
+        ReduceData<Real,Real> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+        for (MFIter mfi(state); mfi.isValid(); ++mfi)
+        {
+            const Box& bx = mfi.fabbox();
+	    auto const& fab = state.array(mfi);
+            reduce_op.eval(bx, reduce_data,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) -> ReduceTuple
+            {
+                Real V_adaptive=0, V_adaptive2=0, V_stupid=0;
+                #include "generated_files/Evolve.cpp_compute_dt_fill"
+                return {V_adaptive, V_stupid};
+            });
+	}
+
+	// extract the reduced values from the combined reduced data structure
+	auto rv = reduce_data.value();
+	Real Vmax_adaptive = amrex::get<0>(rv) + neutrinos.Vvac_max;
+	Real Vmax_stupid   = amrex::get<1>(rv) + neutrinos.Vvac_max;
+
+	// reduce across MPI ranks
+	ParallelDescriptor::ReduceRealMax(Vmax_adaptive);
+	ParallelDescriptor::ReduceRealMax(Vmax_stupid  );
+
+	// define the dt associated with each method
+	Real dt_flavor_adaptive = PhysConst::hbar/Vmax_adaptive*flavor_cfl_factor;
+	Real dt_flavor_stupid   = PhysConst::hbar/Vmax_stupid  *flavor_cfl_factor;
+
+	// pick the appropriate timestep
+	if(dt_flavor_adaptive*max_adaptive_speedup > dt_flavor_stupid)
```

Comment:
Maybe, although in the test cases it seems to work fine, and in a messy multi-D simulation I don't see a situation where that would occur. I don't see it as super important, but if your experience says it would be a good idea I would be happy to put something in.

Reply:
I don't have anything specific in mind, I'm just nervous about the possibility (in code at least) that the timestep could increase by 10x or more between steps?

Reply:
I think the only way this can happen (in our current application) is if there is a large neutrino potential that just happens to cancel, which is what the "stupid" method protects against. The "stupid" limiter also protects against this, so they would have somewhat conflicting roles. I think the real elegant solution would be an error-based adaptive time stepper to replace both.

Reply:
ah, I see, okay!

I'll make an issue for that :)